### DISCUSSION TITLE
Bind FHRP IP to a device

### DISCUSSION THREAD

**Discussion by dasfliege** [netbox-community/netbox#11649] [Q&A]
I am evaluating network documentation tools for my company and i am pretty pleased with what netbox has to offer so far.
But i am stuck a little bit with trying to get all our information represented correctly. Especially with FHRP.

On a network with two sites and full redundancy, we usually have one VRF for each segment. Each VRF contains three VLAN's/Prefixes. One for devices in site A, one for devices in site B and one for the Transit to the router/firewall. All those VRF are present on the switches in site A and in site B. Each switch hosts a dedicated IP for each VLAN, while they share a common VRRP IP.

So my problem is: While it's super easy to create FHRP groups and adding all the IPs involved, i am not able to map some of the IPs to specific devices. An IP can only be mapped to either an FHRP group OR a device, but no to both. But in every VRRP, there are IPs bound to each participating device.

What am i overseeing?

**Comment by jeremystretch** [comment]
In a typical FHRP group setup with two routers providing a gateway VIP to a common segment, there are three IP addresses involved. Using 192.0.2.0/24 as an example, this might be:

- 192.0.2.1 - Assigned to FHRP group
- 192.0.2.2 - Assigned to router A interface 1
- 192.0.2.3 - Assigned to router B interface 1

All three of these IPs should belong to the same VRF (or to the global table). If you have multiple VRFs, just replicate this setup for each VRF. Hope that helps!

**Reply by dasfliege** [reply]
Yeah exactly. But in your example and in my logic, 192.0.2.2 and 192.0.2.3 should be mappable to the FHRP as well which is not possible. If i open a FHRP group it would be nice to not only see the VRRP IP but also all the devices with their dedicated router IPs that are participating in this particular VRRP group.

If i map the router IPs to a manually created VLAN interface on each switch, i see the device as "Member" in the FHRP details page, but it is not showing the IP, which would be a useful information.

I hoped it would work like this:

- I create a FHRP group
- I create a VRRP IP for that group
- I create all the IPs for every participating device/router as role"router" or something, define the priority and map those to a specific device
- Netbox automatically creates an L3 VLAN interface on the mapped device if it doesn't exist already.

That would be quite handy :-)

**Reply by tfboy** [reply]
I'm struggling too to understand how this works.

I have quite a few instances where I need to have one VIP that is tied to two or more other IPs on same the subnet linked to interfaces on devices.

This can be either true network devices, or things like storage (Dell/EMC/Isilon with SSIPs) which have bespoke load balancing and failover where they are accessed by a given IP, for example 192.0.2.1 and then, traffic is automatically redirected to the real interface's IP (192.0.2.2 or 192.0.2.3) depending on load, availability, etc.

The issue I see at the moment is there's no way you can link the FHRP group VIP to the devices' IPs, apart from the fact they're on the same subnet in the same VRF. But in my case, I have a hundred other devices on the same subnet which have nothing to do with the FHRP group so this "link" isn't actually meaningful.

**Reply by tfboy** [reply]
Maybe I'm just missing something.
The documentation refers to FHRP group assignments: https://demo.netbox.dev/static/docs/models/ipam/fhrpgroupassignment/ but I can't see how this is accessed.
Just like there's the "members" table section under FHRP groups: https://demo.netbox.dev/ipam/fhrp-groups/1/ but I can't see how to add a device's or VM's interface to become a member.

**Reply by candlerb** [reply]
> The issue I see at the moment is there's no way you can link the FHRP group VIP to the devices' IPs, apart from the fact they're on the same subnet in the same VRF

They're also on the same *interface* - that's what links them to the device.

Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1
Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1

If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.

> The documentation refers to FHRP group assignments:

The user interface is decidedly non-obvious here. The old way, which still works, is:
* Navigate to a particular device
* *View* one interface, by clicking the interface *name*. (Don't click on the pencil which would *edit* the interface)
* When you're viewing the interface, you'll find a pane with FHRP groups, and there is a button to assign an FHRP group.

The new, alternative way is to click the "+" button next to the interface, and you can go straight to assigning an FHRP group.

![image](https://user-images.githubusercontent.com/44789/229612290-e014bb12-8e4d-46f9-93f1-435fcdbd9b86.png)


See #8106, #10039

**Reply by tfboy** [reply]
> > The issue I see at the moment is there's no way you can link the FHRP group VIP to the devices' IPs, apart from the fact they're on the same subnet in the same VRF
> 
> They're also on the same _interface_ - that's what links them to the device.
> 
> Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1 Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1
> 
> If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.
> 
> > The documentation refers to FHRP group assignments:
> 
> The user interface is decidedly non-obvious here. The old way, which still works, is:
> 
> * Navigate to a particular device
> * _View_ one interface, by clicking the interface _name_. (Don't click on the pencil which would _edit_ the interface)
> * When you're viewing the interface, you'll find a pane with FHRP groups, and there is a button to assign an FHRP group.
> 
> The new, alternative way is to click the "+" button next to the interface, and you can go straight to assigning an FHRP group.
> 
> ![image](https://user-images.githubusercontent.com/44789/229612290-e014bb12-8e4d-46f9-93f1-435fcdbd9b86.png)
> 
> See #8106, #10039

Thank you Brian. It is a bit convoluted, but I've figured it out and have something that can make sense :)

**Reply by dasfliege** [reply]
> They're also on the same interface - that's what links them to the device.
> Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1 Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1
> If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.

I'm still struggling with this. Maybe i'm blind, but i don't think that there is any way to exactly say to which device 192.0.2.2 or 192.0.2.3 belongs in your example. Those are static router IPs, so it has to be possible to map them to a device! All i can do, is creating an interface on the device and map it to the corresponding FHRP group. So i know that one of the IPs contained in the FHRP group is the one that is staticaly set on this device. But i have no chance to find out which one it is exactly. Correct my if i'm wrong, but this is just a pretty relevant information which netbox is not able to reflect right now.

If most of you agree with me, i would consider raising a feature request. A VRRP address must be assignable to a FHRP group AND a device. There is now way around that to have a correct documentation.

**Reply by candlerb** [reply]
> I'm still struggling with this. Maybe i'm blind, but i don't think that there is any way to exactly say to which device 192.0.2.2 or 192.0.2.3 belongs in your example. Those are static router IPs, so it has to be possible to map them to a device!

Of course.

* On device 1, you attach 192.0.2.2 directly to interface d1-X
* On device 2, you attach 192.0.2.3 directly to interface d2-X
* You create FHRP group with IP address 192.0.2.1
* On device 1 and 2, you attach this FHRP group to both d1-X and d2-X

If you search for 192.0.2.3, you'll find device2 interface d2-X only.

If you search for 192.0.2.1, you'll find the FHRP group, and in turn you'll see it linked to both d1-X and d2-X.

> A VRRP address must be assignable to a FHRP group AND a device

No. The VRRP address is assigned to the FHRP group, and the FHRP group is assigned to (one or more) device interfaces.

**Reply by dasfliege** [reply]
Ok got it. So maybe the thing that irritates me is, that under the "members" section in the FHRP overview, there is no column for the IP Address. Think that would be a nice addition in order to quickly see all involved IPs

![image](https://github.com/netbox-community/netbox/assets/31273000/9e009f54-44c5-4466-a8b9-99f82d18954e)

I guess it's correct that the role of the FHRP IP is "VIP" while the IPs on the devices belong to role "VRRP"?

**Reply by candlerb** [reply]
The IP address(es) assigned to the FHRP group itself are shown in the top half of the page, and the device names/interfaces that the FHRP group are assigned to are shown in the bottom half.

Therefore, I think you're asking to see some other IP addresses, not directly related to the FHRP group. These might be:
1. The primary IP address of each device (parent)
2. All other IP addresses assigned to the same interface (in general there could be more than one)

I'm guessing that what you want is (2).  It is one of the columns available in the normal interface table (the interfaces tab on device).

I guess the fundamental problem here is that it's not possible to customize which columns are used in this view, so there would have to be general agreement as to which columns should or should not be present here.  (What about description? label? link peer?)

**Reply by dasfliege** [reply]
I would expect to see the IP that is assigned to the VLAN interface on that device which is participating in the FHRP/VRRP. 

So in your example: 
- "Virtual IP Addresses" should show 192.0.2.1 which is the VIP of the VRRP
- "Members" should show Router1 with IP 192.0.2.2 on Interface xxxxxx
- "Members" should show Router2 with IP 192.0.2.3 on Interface xxxxxx

I agree with you that the main problem seems to be, that the columns aren't customizable. Also if there are several IPs on that specific interface, the additional column should either show all IPs or there needs to be a separate property for an interface/IP that allows you to choose which IP is the primary IP for the interface. I think right now it is only possibly to declare an IP as primary IP for the device, but not for the interface.

**Comment by mtinberg** [comment]
I think the way it works is that if you were templating the config for router A interface 1 (/dcim/interfaces/12345/) you'd see the related IPs 192.0.2.1 and 192.0.2.2 but not 192.0.2.3 which is on router B, router B interface 1 (/dcim/interfaces/23456/) would be related to 192.0.2.1 and 192.0.2.3, from the IP record you can see that one is assigned to a group/cluster/HA and the other directly assigned to a device/interface.

  *   192.0.2.1 - Assigned to FHRP group
  *   192.0.2.2 - Assigned to router A interface 1
  *   192.0.2.3 - Assigned to router B interface 1

> netbox automatically creates an L3 VLAN interface on the mapped device if it doesn't exist already.

This kind of logic would be better in a Custom Script which can be coded for your particular kind of device and rules about how it should work, there is nothing built-in Netbox which is going to create an Interface database record automatically, those are either created from the Device Type/Module Type templates (with some ability to template a slot number/stack number into the interface name), manually through the UI or using a Script or API client.

If you wanted to make a Script where you could pick a Prefix and two Devices and have it auto-allocate the IPs, create the FHRP group, create the Interfaces and associate them all together then that could solve the problem the way you want.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by dasfliege** [reply]
Well i think i understand how it is meant to be in netbox, but it does not makes much sense for me. I really think it should be possible to assign an IP objects to a FHRP group AND a device. That would make it much more transparent.
Or at least the router IP should be one of the columns displayed in the "Members" section:

![image](https://user-images.githubusercontent.com/31273000/216464088-5ecdf4a3-0e09-480a-8787-c01883f023d1.png)

**Comment by dasfliege** [comment]
> > The issue I see at the moment is there's no way you can link the FHRP group VIP to the devices' IPs, apart from the fact they're on the same subnet in the same VRF
> 
> They're also on the same _interface_ - that's what links them to the device.
> 
> Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1 Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1
> 
> If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.
> 
> > The documentation refers to FHRP group assignments:
> 
> The user interface is decidedly non-obvious here. The old way, which still works, is:
> 
> * Navigate to a particular device
> * _View_ one interface, by clicking the interface _name_. (Don't click on the pencil which would _edit_ the interface)
> * When you're viewing the interface, you'll find a pane with FHRP groups, and there is a button to assign an FHRP group.
> 
> The new, alternative way is to click the "+" button next to the interface, and you can go straight to assigning an FHRP group.
> 
> ![image](https://user-images.githubusercontent.com/44789/229612290-e014bb12-8e4d-46f9-93f1-435fcdbd9b86.png)
> 
> See #8106, #10039



> They're also on the same _interface_ - that's what links them to the device.
> 
> Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1 Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1
> 
> If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.



> They're also on the same _interface_ - that's what links them to the device.
> 
> Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1 Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1
> 
> If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.



> They're also on the same _interface_ - that's what links them to the device.
> 
> Device 1 interface X: IP address 192.0.2.2, FHRP 192.0.2.1 Device 2 interface Y: IP address 192.0.2.3, FHRP 192.0.2.1
> 
> If you search for the FHRP address 192.0.2.1, you'll immediately see what interfaces it's linked to - although you have to click through to find out the other IPs on those interfaces.

**Comment by lukedaemon** [comment]
Hello guys,
after reading your interesting conversation, I'll have a question for you: assuming there are two devices (cluster); each device has multiple interfaces, one for each VLAN; and each VLAN has a VIP. Do you suggest creating one FHRP group with all VIPs and all interfaces (both devices) as members, or one FHRP group for each VLAN ?

**Reply by candlerb** [reply]
It depends what you have implemented.

In the normal case, I would expect separate instances of VRRP or HSRP on each VLAN, which can be failed over individually (and/or have different priorities on each VLAN). In that case I'd use separate FHRP groups in Netbox.

If you have built something where somehow all the VIPs are linked together and failover together, then I guess you could use a single FHRP group. But it would be a bit awkward, because the FHRP group would have multiple IP addresses, and you would associate the entire FHRP with all the VLAN interfaces, but in reality each IP address only belongs to *one* VLAN.  So I think you'd need a good reason for doing it that way.

**Comment by lukedaemon** [comment]
Hello @candlerb,
yep all VIPs failover together; however I'll follow your suggestion. Having one FHRP group for each VLAN (even if they are treated in the same manner during a failover) will result in a more clear documentation.
